src/uct/ib/mlx5/dc/dc_mlx5.c

```diff
@@ -267,14 +267,27 @@ static void uct_dc_mlx5_iface_progress_enable(uct_iface_h tl_iface, unsigned fla
     uct_base_iface_progress_enable_cb(&iface->super.super, iface->progress, flags);
 }
 
+static void uct_dc_mlx5_asan_relocate_dcis_array(uct_dc_mlx5_iface_t *iface)
+{
+#if __SANITIZE_ADDRESS__
+    size_t buffer_size = sizeof(uct_dc_dci_t) * ucs_array_capacity(&iface->tx.dcis);
+    size_t num_dcis = ucs_array_length(&iface->tx.dcis);
+    void *old_buffer_p;
+
+    ucs_debug_asan_relocate_array_buffer((void**)&iface->tx.dcis.buffer,
+                                         &old_buffer_p, buffer_size);
+    uct_dc_mlx5_iface_dcis_array_copy(iface->tx.dcis.buffer, old_buffer_p,
+                                      num_dcis);
+    ucs_array_buffer_free(old_buffer_p);
+#endif
+}
+
 static UCS_F_ALWAYS_INLINE unsigned
 uct_dc_mlx5_poll_tx(uct_dc_mlx5_iface_t *iface, int poll_flags)
```

Comment:
Is it enough to fix only code in uct_dc_mlx5_poll_tx?
I see that `uct_dc_mlx5_iface_resize_and_fill_dcis` may be called directly/indirectly from:
- init functions, hopefully ok: uct_dc_mlx5_ep_basic_init, uct_dc_mlx5_ep_basic_init, uct_dc_mlx5_iface_init_dcis_array
- uct_dc_mlx5_iface_dci_can_alloc_or_create -> uct_dc_mlx5_iface_dci_get, which is called from many places:
1. UCT_DC_MLX5_CHECK_DCI_RES
uct_dc_mlx5_ep_fc_pure_grant_send
uct_dc_mlx5_ep_fc_hard_req_send
UCT_DC_MLX5_CHECK_RES - MANY references!
3. UCT_DC_CHECK_RES_PTR

Basically this function may be called indirectly from dozens of different flows, and it's hard to track all of them.. And each workflow caches DCI internals (txqp, txwq) into local variables, so I'm not sure we can easily guarantee that it always works reliably. Even if we fix it now, it is very error-prone and it's quite hard to find the root cause

One option to make it 100% safe is to allocate TX structures on heap:
```C
typedef struct uct_dc_dci {
    uct_rc_txqp_t                *txqp; /* DCI qp */
    uct_ib_mlx5_txwq_t           *txwq; /* DCI mlx5 wq */
```
By doing this all the other problems are solved:
1. DCI object becomes "primitive" in the sense we don't even need advanced copy logic.
2. It's safe to cache those fields from any workflow, no need to access by index
I think it has almost zero overhead in terms of performance

Reply:
@iyastreb it is indeed error-prone, though your suggestion would introduce a performance penalty on fast path.

Reply:
@roiedanino Any comments on that?

Reply:
After reconsidering that -
@roiedanino , let's go with Ilia's suggestion.